### PR TITLE
Feature/vmo 6022/update block exits part2

### PR DESCRIPTION
**Pull request by safydy-r**
**Change summary:**
- Instead of customizing exits number & test in store level, I moved it into `config` level
- That will make viamo customization more cleaner, because some blocks are having 01 exit in community version but 02 exits in viamo (setContact & Output)

**Overview**
Screen Shot 2022-07-18 at 4 39 51 PM

### CODE REVIEW

File: src/store/flow/block-types/BaseBlock.ts [hunk 1]
```diff
@@ -27,13 +26,9 @@ export const actions = {
...
       semantic_label: '',
       config: {},
-      exits: props?.exits ?? [
-        await dispatch('flow/block_createBlockDefaultExitWith', {
-          props: ({
-            uuid: await (new IdGeneratorUuidV4()).generate(),
-          }) as IBlockExit,
-        }, {root: true}),
-      ],
+      exits: props?.exits ?? await dispatch('flow/block_generateExitsBasedOnUiConfig', {
```
Comment (safydy-r) [inline src/store/flow/block-types/BaseBlock.ts:29]: I removed all exit generation within each block's `creatWith` in favour of this new line.

File: src/store/flow/block-types/BaseBlock.ts [hunk 2]
```diff
@@ -51,16 +46,22 @@ export const actions = {
...
   async handleBranchingTypeChangedToUnified(
-    {dispatch}: {dispatch: Dispatch},
+    {dispatch, rootGetters}: {dispatch: Dispatch, rootGetters: any},
     {block}: {block: IBlock},
   ): Promise<void> {
-    return dispatch('flow/block_updateBranchingExitsWithInvalidScenario', {
-      blockId: block.uuid,
-      test: 'block.value = true',
-    }, {root: true})
+    if (rootGetters['flow/block_shouldHave2Exits'](block.type) === true) {
```
Comment (safydy-r) [inline src/store/flow/block-types/BaseBlock.ts:56]: The current logic will now handle all blocks'  handleBranchingTypeChangedToUnified, apart from MCQ which could use dynamic test values.
Therefore, I removed all handleBranchingTypeChangedToUnified from each block's store ;)

File: src/store/flow/block.ts [hunk 3]
```diff
@@ -263,15 +310,20 @@ export const actions: ActionTree<IFlowsState, IRootState> = {
   /**
    * Update exits after the block creation.
    * Standard exit where the end user can make an error. We have 02 exits:
-   * - Valid
-   * - Invalid
+   * - Non default exits would be collapsed into 01 exit (eg: Valid, Success)
+   * - Default exit would be the last exit (eg: Invalid, Failure)
    */
-  async block_updateBranchingExitsWithInvalidScenario({state, dispatch}, {blockId, test}: {blockId: IBlock['uuid'], test: IBlockExit['test']}) {
+  async block_resetBranchingExitsByCollapsingNonDefault(
```
Comment (safydy-r) [inline src/store/flow/block.ts:316]: renamed with `reset` to clarify the intension

File: src/store/flow/block-types/BaseBlock.ts [hunk 4]
```diff
@@ -48,19 +48,36 @@ export const actions = {
       vendor_metadata: await dispatch('initiateExtraVendorConfig'),
     },
     )
+
+    // Define exits after we have the whole final props, this is important for dynamic test value
+    if (props?.exits === undefined) {
```
Comment (safydy-r) [inline src/store/flow/block-types/BaseBlock.ts:53]: I removed all exit generation within each block's `creatWith()` in favour of this new block of codes

File: src/store/flow/block-types/BaseBlock.ts [hunk 5]
```diff
@@ -48,19 +48,36 @@ export const actions = {
...
+
+    return mainProps
   },
 
   /**
-   * This will be the default standard exit mode, but we can override it in the specific block store
+   * This will be the default standard exit mode,
+   * but we can override it in the specific block store (eg: for dynamic test generation in MCQ)
    */
   async handleBranchingTypeChangedToUnified(
```
Comment (safydy-r) [inline src/store/flow/block-types/BaseBlock.ts:67]: Edited this to make it more generic for blocks with 2 or 1 exits

File: src/store/flow/block-types/BaseBlock.ts [hunk 6]
```diff
@@ -1,23 +1,30 @@
...
+   */
+  primaryExitTest: () => (_blockProps: Partial<IBlock>) => undefined,
+}
 
 export const mutations: MutationTree<IEmptyState> = {}
 
 export const actions = {
   async createWith(
-    {dispatch}: {dispatch: Dispatch},
+    {getters, dispatch}: {getters: any, dispatch: Dispatch},
```
[the remark this answers is not in the document]
Comment (bulatgab) [reply src/store/flow/block-types/BaseBlock.ts:24]: You can do `{getters, dispatch}: ActionContext<IEmptyState, IRootState>`

[the remark this answers is not in the document]
Comment (bulatgab) [reply src/store/flow/block-types/BaseBlock.ts:24]: (and the same everywhere else)

[the remark this answers is not in the document]
Comment (safydy-r) [reply src/store/flow/block-types/BaseBlock.ts:24]: This is very helpful, thanks @bulatgab

**Pull request merged**